app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/KiwixWifiP2pBroadcastReceiver.java

```diff
@@ -0,0 +1,67 @@
+package org.kiwix.kiwixmobile.zim_manager.local_file_transfer;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.Intent;
+import android.net.NetworkInfo;
+import android.net.wifi.p2p.WifiP2pDevice;
+import android.net.wifi.p2p.WifiP2pManager;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+/**
+ * Helper class for the local file sharing module.
+ *
+ * Handles the broadcasts pertaining to the wifi p2p group formed in WiFi Direct. Works along with
+ * the wifi p2p manager in {@link LocalFileTransferActivity}.
+ */
+public class KiwixWifiP2pBroadcastReceiver extends BroadcastReceiver {
+
+  private P2pEventListener p2pEventListener;
+
+  public KiwixWifiP2pBroadcastReceiver(@NonNull P2pEventListener p2pEventListener) {
+    this.p2pEventListener = p2pEventListener;
+  }
+
+  @Override
+  public void onReceive(@NonNull Context context, @NonNull Intent intent) {
+
+    switch(intent.getAction()) {
+      case WifiP2pManager.WIFI_P2P_STATE_CHANGED_ACTION: {
+        int wifiP2pState = intent.getIntExtra(WifiP2pManager.EXTRA_WIFI_STATE, -1);
+        p2pEventListener.onWifiP2pStateChanged(wifiP2pState);
+        break;
+      }
+
+      case WifiP2pManager.WIFI_P2P_PEERS_CHANGED_ACTION: {
+        p2pEventListener.onPeersChanged();
+        break;
+      }
+
+      case WifiP2pManager.WIFI_P2P_CONNECTION_CHANGED_ACTION: {
+        NetworkInfo networkInfo = intent.getParcelableExtra(WifiP2pManager.EXTRA_NETWORK_INFO);
```

Comment:
I would also keep the networkInfo logic internal to the broadcastReceiver instead of putting it in WDM

Reply:
The NetworkInfo object is only used to check whether the device is connected or not. Since WDM is handling all the logic related to using wifi p2p, shouldn't it be passed on the whole object and respond as necessary (instead of just the state of connection)?

(That is also kind of what I'm doing with the remaining callbacks - KiwixBroadcastReceiver extracts the relevant object from the broadcast intent and forwards it to WDM, which handles the response to the broadcast on its own)

Reply:
It is just a lot purer if you call methods that are events, KiwixBroadcastReceiver should help make your life easier when you use it but now if you ever use it again you have to perform logic in the client when what you actually care about is `what happened` not `how did it happen`.

I also generally don't like passing back Android classes. Intents are just the worst container of data in the world and having your own class makes things nicer and allows you to give names with actual context. If you look at ConnectivityBroadcastReceiver I was returning an enum because it was much easier to read at the call site what was going on as opposed to a Boolean (and allows me to add more states in the future if I want)

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/1baafc9db4270b2ccf492be740c443791b01ae2b#diff-e84e03042026401a0660fa610753ae45L63)